include/deal.II/lac/computation.h

```diff
@@ -0,0 +1,639 @@
+ * to <code>A</code>, <code>b</code> and <code>x</code> and defers the
+ * actual computation until <code>result</code>, or <code>result_add</code>
+ * are explicitly invoked,
+ * @code
+ *   dealii::Vector<double> y;
+ *   residual.reinit_vector(y);
+ *   residual.result(y);
+ *   residual.result_add(y);
+ * @endcode
+ * or until the @p Computation object is implictly converted:
+ * @code
+ *   dealii::Vector<double> y;
+ *   y = residual;
+ *   y += residual;
+ *   y -= residual;
+ * @endcode
+ *
+ * @author Matthias Maier, 2015
+ *
+ * @ingroup LAOperators
+ */
+template <typename Range> class Computation
+{
+public:
+
+  /**
+   * Create an empty Computation object. All <code>std::function</code>
+   * member objects are initialized with default variants that throw an
+   * exception upon invocation.
+   */
+  Computation ()
+  {
+    result = [](Range &)
+    {
+      Assert(false,
+             ExcMessage("Uninitialized Computation<Range>::result called"));
+    };
+
+    result_add = [](Range &)
+    {
+      Assert(false,
+             ExcMessage("Uninitialized Computation<Range>::result_add called"));
+    };
+
+    reinit_vector = [](Range &, bool)
+    {
+      Assert(false,
+             ExcMessage("Uninitialized Computation<Range>::reinit_vector "
+                        "method called"));
+    };
+  }
+
+  /**
+   * Default copy constructor.
+   */
+  Computation (const Computation<Range> &) = default;
+
+  /**
+   * Constructor that creates a Computation object from a reference vector
+   * @p u. The Computation returns @p u.
+   *
+   * The Computation object that is created stores a reference to @p u.
+   * Thus, the vector must remain a valid reference for the whole lifetime
+   * of the Computation object. All changes made on @p u after the creation
+   * of the Computation object are reflected by the operator object.
+   */
+  Computation (const Range &u)
+  {
+    *this = u;
+  }
+
+  /**
+   * Default copy assignment operator.
+   */
+  Computation<Range> &operator=(const Computation<Range> &) = default;
+
+  /**
+   * Copy assignment operator that creates a Computation object from a
+   * reference vector @p u. The Computation returns @p u.
+   *
+   * The Computation object that is created stores a reference to @p u.
+   * Thus, the vector must remain a valid reference for the whole lifetime
+   * of the Computation object. All changes made on @p u after the creation
+   * of the Computation object are reflected by the operator object.
+   */
+  Computation<Range> &operator=(const Range &u)
+  {
+    result = [&u](Range &v)
+    {
+      v = u;
+    };
+
+    result_add = [&u](Range &v)
+    {
+      v += u;
+    };
+
+    reinit_vector = [&u](Range &v, bool fast)
+    {
+      v.reinit(u, fast);
+    };
+
+    return *this;
+  }
+
+  /**
+   * Convert a Computation to its result.
+   *
+   * This conversion operator creates a vector of the Range space and calls
+   * <code>result()</code> on it.
+   */
```

Comment:
> but not of the allocation and deallocation of memory for the temporary

No. Which temporary are you talking about?

Because at the point when the conversion operator is called, you actually want to have a vector. It has to be created eventually. There is no (performance) difference between:
```
Vector<double> foo;
[...].PackedOperation::reinit_vector(foo);
[...]PackedOperation::result(foo);
```
and
```
Vector<double> foo = bar;
```
Further, with move assignment operators, also a subsequent ```foo = bar``` involves no temporary.